nominal/sdk.py

```diff
@@ -0,0 +1,508 @@
+from __future__ import annotations
+
+from io import TextIOBase
+import time
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta
+from types import MappingProxyType
+from typing import BinaryIO, Iterable, Literal, Mapping, Sequence, cast
+
+import certifi
+from conjure_python_client import RequestsClient, ServiceConfiguration, SslConfiguration
+
+from ._api.combined import attachments_api
+from ._api.combined import scout_catalog
+from ._api.combined import scout
+from ._api.combined import scout_run_api
+from ._api.ingest import ingest_api
+from ._api.ingest import upload_api
+from ._multipart import put_multipart_upload
+from ._utils import (
+    _conjure_time_to_integral_nanoseconds,
+    _flexible_time_to_conjure_scout_run_api,
+    _timestamp_type_to_conjure_ingest_api,
+    construct_user_agent_string,
+    CustomTimestampFormat,
+    IntegralNanosecondsUTC,
+    Self,
+    TimestampColumnType,
+)
+from .exceptions import NominalIngestError, NominalIngestFailed
+
+_AllowedFileExtensions = Literal[".csv", ".csv.gz", ".parquet"]
+
+__all__ = [
+    "NominalClient",
+    "Run",
+    "Dataset",
+    "Attachment",
+    "IntegralNanosecondsUTC",
+    "CustomTimestampFormat",
+    "NominalIngestError",
+    "NominalIngestFailed",
+]
+
+
+@dataclass(frozen=True)
+class Run:
+    rid: str
+    title: str
+    description: str
+    properties: Mapping[str, str]
+    labels: Sequence[str]
+    start: IntegralNanosecondsUTC
+    end: IntegralNanosecondsUTC | None
+    _auth_header: str = field(repr=False)
+    _run_client: scout.RunService = field(repr=False)
+
+    def add_datasets(self, datasets: Mapping[str, str]) -> None:
```

Comment:
thoughts on this supporting `[str, Union[str, Dataset]]` instead to make it more intuitive to use with existing objects?

Reply:
good idea

Reply:
can do that for all the method args that take RIDs

Reply:
done. I also added it to `get_<x>()` methods, which is maybe a bit silly, e.g. `attachment = get_attachment(attachment)`, but figure I might as well keep the API uniform.